### PR TITLE
Revert all PR for "Data providers perform unnecessary COUNT queries that negatively affect performance".

### PR DESCRIPTION
**Pull request by terabytesoftw**
PRs Revert:

- https://github.com/yiisoft/yii2/pull/20223 
- https://github.com/yiisoft/yii2/pull/20222
- https://github.com/yiisoft/yii2/pull/20225
- https://github.com/yiisoft/yii2/pull/20007

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | https://github.com/yiisoft/yii2/issues/20213

### CODE REVIEW

File: framework/data/ActiveDataProvider.php [hunk 1]
```diff
@@ -159,8 +166,13 @@ protected function prepareTotalCount()
         if (!$this->query instanceof QueryInterface) {
             throw new InvalidConfigException('The "query" property must be an instance of a class that implements the QueryInterface e.g. yii\db\Query or its subclasses.');
         }
-        $query = clone $this->query;
-        return (int) $query->limit(-1)->offset(-1)->orderBy([])->count('*', $this->db);
+        $query = (clone $this->query)->limit(-1)->offset(-1)->orderBy([]);
+        $key = md5((string)$query);
```
Comment (sleptor) [inline framework/data/ActiveDataProvider.php:170]: IMHO, query object serialization is unreliable. For example, logically identical queries produce different hashes.
```
X::find()->andWhere(['id' => 1, 'status' => 1]);
X::find()->andWhere(['status' => 1, 'id' => 1]);
```
Can we simply cache the value and add a kind of flush() function for the cases when we need to recalculate the counter?

Comment (terabytesoftw) [reply framework/data/ActiveDataProvider.php:170]: You can do that in a separate PR, here we are only going to revert the changes, now you could test this branch,

Thanks you.

Comment (sleptor) [reply framework/data/ActiveDataProvider.php:170]: Don't we revert to 2.0.49.x state? There is no query cache at all
https://github.com/yiisoft/yii2/blob/2.0.49.x/framework/data/ActiveDataProvider.php#L168

Comment (terabytesoftw) [reply framework/data/ActiveDataProvider.php]: Thks, fixed.

File: framework/data/Pagination.php [hunk 2]
```diff
@@ -262,7 +264,7 @@ public function createUrl($page, $pageSize = null, $absolute = false)
             $request = Yii::$app->getRequest();
             $params = $request instanceof Request ? $request->getQueryParams() : [];
         }
-        if ($page > 0 || ($page === 0 && $this->forcePageParam)) {
+        if ($page > 0 || $page == 0 && $this->forcePageParam) {
```
[the remark this answers is not in the document]
Comment (marcovtwout) [reply framework/data/Pagination.php:267]: I don't think we should refactor anything in this PR: just cleanly revert the changes first.

**Comment by terabytesoftw** [comment]
> @terabytesoftw The 4 lines mensioned in the opening posts are exactly the ones that need to be removed, but it looks like this PR is currently incomplete? I don't see all the expected changes in this PR (except for the CHANGELOG entries for 2.0.50 that should not be reverted)

Could you explain to me what else is missing, i reverted all the PRs, compared with the `2.0.49.x` branch and i don't see any differences.

**Comment by marcovtwout** [comment]
@terabytesoftw It's not easy to read from all the separate diffs. Here's the things I noticed manually:

- the removed `use Yii` statements in many files from https://github.com/yiisoft/yii2/pull/20222/files#diff-2d59a53b455d32e67467aa1a4036d737a3564d6f65203e405ee3fc8c251787de, or did you leave these removed on purpose?
- one CHANGELOG line can be removed, since 2.0.50.1 is not released yet: https://github.com/yiisoft/yii2/blob/master/framework/CHANGELOG.md?plain=1#L10
- new CHANGELOG line can be added for 2.0.50.1 that details the changes you are making

I agree a final comparison for all relevant files with 2.0.49.x is a good thing (but double check any other relevant changes to those files from other PR's are not reverted).

**Comment by terabytesoftw** [comment]
* the removed `use Yii` statements in many files from https://github.com/yiisoft/yii2/pull/20222/files#diff-2d59a53b455d32e67467aa1a4036d737a3564d6f65203e405ee3fc8c251787de, or did you leave these removed on purpose?

Yes, i removed it, since it is a change that does not affect.

**Comment by samdark** [comment]
Thanks.

**Pull request merged**